app/src/gecko/java/org/mozilla/focus/web/GeckoViewPrompt.java

```diff
@@ -0,0 +1,882 @@
+                        });
+        createStandardDialog(addCheckbox(builder, container, callback), callback).show();
+    }
+
+    private void addChoiceItems(final int type, final ArrayAdapter<GeckoBundle> list,
+                                final GeckoBundle[] items, final String indent) {
+        if (type == CHOICE_TYPE_MENU) {
+            list.addAll(items);
+            return;
+        }
+
+        for (final GeckoBundle item : items) {
+            final GeckoBundle[] children = item.getBundleArray("items");
+            if (indent != null && children == null) {
+                item.putString("label", indent + item.getString("label", ""));
+            }
+            list.add(item);
+
+            if (children != null) {
+                final String newIndent;
+                if (type == CHOICE_TYPE_SINGLE || type == CHOICE_TYPE_MULTIPLE) {
+                    newIndent = (indent != null) ? indent + '\t' : "\t";
+                } else {
+                    newIndent = null;
+                }
+                addChoiceItems(type, list, children, newIndent);
+            }
+        }
+    }
+
+    public void promptForChoice(final GeckoSession session, final String title,
+                                final String msg, final int type,
+                                final GeckoBundle[] choices, final ChoiceCallback callback) {
+        final Activity activity = mActivity;
+        if (activity == null) {
+            callback.dismiss();
+            return;
+        }
+        final AlertDialog.Builder builder = new AlertDialog.Builder(activity);
+        addStandardLayout(builder, title, msg);
+
+        final ListView list = new ListView(builder.getContext());
+        if (type == CHOICE_TYPE_MULTIPLE) {
+            list.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE);
+        }
+
+        final ArrayAdapter<GeckoBundle> adapter = new ArrayAdapter<GeckoBundle>(
+                builder.getContext(), android.R.layout.simple_list_item_1) {
+            private static final int TYPE_MENU_ITEM = 0;
+            private static final int TYPE_MENU_CHECK = 1;
+            private static final int TYPE_SEPARATOR = 2;
+            private static final int TYPE_GROUP = 3;
+            private static final int TYPE_SINGLE = 4;
+            private static final int TYPE_MULTIPLE = 5;
+            private static final int TYPE_COUNT = 6;
+
+            private LayoutInflater mInflater;
+            private View mSeparator;
+
+            @Override
+            public int getViewTypeCount() {
+                return TYPE_COUNT;
+            }
+
+            @Override
+            public int getItemViewType(final int position) {
+                final GeckoBundle item = getItem(position);
+                if (item.getBoolean("separator")) {
+                    return TYPE_SEPARATOR;
+                } else if (type == CHOICE_TYPE_MENU) {
+                    return item.getBoolean("selected") ? TYPE_MENU_CHECK : TYPE_MENU_ITEM;
```

Comment:
\[PMD\] INFO: The String literal "selected" appears 4 times in this file; the first occurrence is on line 324